cpu/sam0_common/periph/eth.c

```diff
@@ -0,0 +1,312 @@
+                        |  (mac->uint8[1] << 8)  | mac->uint8[0]);
+}
+
+void sam0_eth_get_mac(eui48_t *out)
+{
+    out->uint8[5] = (GMAC->Sa[0].SAT.reg >> 8);
+    out->uint8[4] = (GMAC->Sa[0].SAT.reg);
+    out->uint8[3] = (GMAC->Sa[0].SAB.reg >> 24);
+    out->uint8[2] = (GMAC->Sa[0].SAB.reg >> 16);
+    out->uint8[1] = (GMAC->Sa[0].SAB.reg >> 8);
+    out->uint8[0] = (GMAC->Sa[0].SAB.reg);
+}
+
+int sam0_eth_send(const struct iolist *iolist)
+{
+    unsigned len = iolist_size(iolist);
+    unsigned tx_len = 0;
+    tx_curr = &tx_desc[tx_idx];
+
+    /* load packet data into TX buffer */
+    for (const iolist_t *iol = iolist; iol; iol = iol->iol_next) {
+        if (tx_len + iol->iol_len > ETHERNET_MAX_LEN) {
+            return -EBUSY;
+        }
+        if (iol->iol_len) {
+            memcpy ((uint32_t*)(tx_curr->address + tx_len), iol->iol_base, iol->iol_len);
+            tx_len += iol->iol_len;
+        }
+    }
+    if (len == tx_len) {
+        tx_curr->status |= (len & DESC_TX_STATUS_LEN_MASK);
+        tx_curr->status |= DESC_TX_STATUS_LAST_BUF | DESC_TX_STATUS_USED;
+
+        tx_idx = (tx_idx < ETH_TX_BUFFER_COUNT-1) ? tx_idx+1 : 0;
+        __DSB();
+        tx_curr->status &= ~DESC_TX_STATUS_USED;
+        /* Start transmission */
+        GMAC->NCR.reg |= GMAC_NCR_TSTART;
+    }
+    else {
+        DEBUG("Mismatch TX len, abort send\n");
+    }
+    return len;
+}
+
+static int _try_receive(char* data, int max_len, int block)
+{
+    (void)block;
+    unsigned rxlen = 0;
+    uint16_t idx = rx_idx;
+    /* Ensure we are at the beginning of the new frame */
+    while (!(rx_curr->address & DESC_RX_ADDR_OWNSHP) && (rx_curr->status & DESC_RX_STATUS_STA_FRAME)) {}
+
+    for (unsigned cpt=0; cpt < ETH_RX_BUFFER_COUNT; cpt++) {
+        /* Get the length of the received frame */
+        unsigned len = (rx_curr->status & DESC_RX_STATUS_FRAME_LEN_MASK);
+
+        /* Only copy data if the stack requested it, otherwise return the length
+           of the next frame if available */
+        if (max_len) {
+            /* If buffer available, copy data into it */
+            if (data)  {
+                memcpy(&data[rxlen], rx_buf[idx], len);
+            }
+            /* Tell the GMAC IP that we don't need this frame anymore  */
+            rx_curr->address &= ~DESC_RX_ADDR_OWNSHP;
+        }
+
+        rxlen += len;
+
+        if (rx_curr->status & DESC_RX_STATUS_END_FRAME) {
+            /* We reach the end of frame, leave the loop */
+            break;
+        }
+        /* Prepare next buffer */
+        idx = (idx + 1) % ETH_RX_BUFFER_COUNT;
+        rx_curr = &rx_desc[idx];
+
+   }
+   /* restore the previous index if packets were not released */
+   if (!max_len) {
+       rx_curr = &rx_desc[rx_idx];
+   }
+   /* Prepare next buffer if packets were released */
+   else {
+       rx_idx = (idx+1) % ETH_RX_BUFFER_COUNT;
+       rx_curr = &rx_desc[rx_idx];
```

Comment:
But at this point we already copied the content of the current RX buffer to the upper layer, we might as well re-use it (that would make the multiple buffers a bit pointless).
Why not do this at the beginning? 

Or can the hardware switch those buffers on it's own?

Reply:
Hardware have its own index and it will use the next available buffer anyways so we'd better do it too to avoid iterating too much to check which buffer should we use.
GMAC IP will parse the status field of each available buffer descriptor and it will take the ownership of a buffer by setting the rightful bit.

> Why not do this at the beginning?

What do you means ?

Reply:
> Hardware have its own index and it will use the next available buffer anyways so we'd better do it too to avoid iterating too much to check which buffer should we use.

ah that clears it up.
How do we know we are in sync with the hardware? Can we find out what buffer it is currently at, without keeping track manually?